compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -640,6 +640,34 @@ impl<'a> Parser<'a> {
             }
         }
 
+        // Extra info for `c"str"` before 2021 edition or `c "str"` in all editions. The heuristic
+        // here is to identify a cooked, uninterpolated `c` id immediately followed by a string, or
+        // a cooked, uninterpolated `cr` id immediately followed by a string or a `#`, in an
+        // edition where c-string literals are not allowed. There is the slight possibility of a
+        // false positive for a `cr#` that wasn't intended to start a c-string literal, but the
+        // lexer was greedy and didn't preserve whether the `r#` on its own would have started a
+        // valid raw string literal.
+        if ((self.prev_token.kind == TokenKind::Ident(sym::c, false)
+            && matches!(&self.token.kind, TokenKind::Literal(token::Lit { kind: token::Str, .. })))
+            || (self.prev_token.kind == TokenKind::Ident(sym::cr, false)
+                && matches!(
+                    &self.token.kind,
+                    TokenKind::Literal(token::Lit { kind: token::Str, .. }) | token::Pound
+                )))
+            && self.prev_token.span.hi() == self.token.span.lo()
+            && !self.token.span.at_least_rust_2021()
+        {
+            err.cancel();
+            let descr = super::token_descr(&self.token);
+            let mut err =
+                self.struct_span_err(self.token.span, format!("found unexpected token {descr}"));
+            err.span_label(self.token.span, "found here".to_owned());
+            err.note("you may be trying to declare a c-string literal");
```

Comment:
```suggestion
            err.note("you might have meant to write a c-string literal");
```

Not a native speaker but the word *declare* doesn't feel right to me in this context.